src/prefect/events/related.py

```diff
@@ -0,0 +1,94 @@
+from typing import TYPE_CHECKING, Dict, List, Optional, Set, Tuple
+from uuid import UUID
+
+from .schemas import RelatedResource
+
+if TYPE_CHECKING:
+    from prefect.client.schemas import FlowRun
+    from prefect.server.schemas.core import Flow
+
+related_resource_cache: Dict[UUID, Tuple[Optional["FlowRun"], Optional["Flow"]]] = {}
+
+
+async def related_resources_from_run_context(
+    exclude: Optional[Set[str]] = None,
+) -> List[RelatedResource]:
+    from prefect.context import FlowRunContext, TaskRunContext
+
+    if exclude is None:
+        exclude = set()
+
+    flow_run_id: Optional[UUID] = None
+    flow_run: Optional["FlowRun"] = None
+    flow: Optional["Flow"] = None
+
+    flow_run_context = FlowRunContext.get()
+
+    if flow_run_context is None:
+        task_run_context = TaskRunContext.get()
```

Comment:
Generally we have the task run context take precedence over the flow run context — not the other way around. If you emit an event from a task in a flow without a remote task runner then this looks like it will get skipped?